source/dotnet/Library/AdaptiveCards/AdaptiveFeatureRegistration.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AdaptiveCards
+{
+    public class AdaptiveFeatureRegistration
+    {
+        public AdaptiveFeatureRegistration()
+        {
+            features[adaptiveCardFeature] = "1.2";
+        }
+
+        public string Get(string feature)
+        {
+            if (!features.ContainsKey(feature))
+            {
+                return "";
+            }
+            else
+            {
+                return features[feature];
+            }
+        }
+
+        public void Set(string feature, string version)
+        {
+            if (feature == adaptiveCardFeature)
+            {
+                throw new AdaptiveException("Not allowed to override adaptiveCards version");
+            }
+            else
+            {
+                features[feature] = version;
```

Comment:
>    features[feature] = version; [](start = 12, length = 32)

WHat happens if you try to add the same feature twice?

Reply:
I'm a little on the fence about this one... As it stands, here's are the classes of things that we warn on:

```
    enum class WarningStatusCode
    {
        UnknownElementType = 0,
        UnknownActionElementType,
        UnknownPropertyOnElement,
        UnknownEnumValue,
        NoRendererForType,
        InteractivityNotSupported,
        MaxActionsExceeded,
        AssetLoadFailed,
        UnsupportedSchemaVersion,
        UnsupportedMediaType,
        InvalidMediaMix,
        InvalidColorFormat,
        InvalidDimensionSpecified,
        InvalidLanguage,
    };
```

I could add another one of course, but I'm not sure this would rise to the same level. It's probably worth noting that this is something that app authors would set on a renderer before rendering a parsed card. It's not unsupported, but it's probably a little unusual for folks to reuse a renderer. Certainly it would feel a bit odd to reuse a renderer while swapping out versions of a supported feature.

---
In reply to: [278322077](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278322077) [](ancestors = 278322077)

Reply:
That's fine - i also noticed the API is called "Set" which shuld indicate "addorupdate" behavior by its name. (it's not for eg called "Add/Insert") so should be fine -- 

---
In reply to: [278341499](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278341499) [](ancestors = 278341499,278322077)